ShardedTensor framework for ChunkedShardingSpec

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58517 ShardedTensor framework for ChunkedShardingSpec**

Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

pritamdamania committed:
ShardedTensor framework for ChunkedShardingSpec

Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]

divchenko commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -0,0 +1,186 @@
+        """
+        return self._sharding_metadata
+
+    def local_shards(self) -> torch.Tensor:
```
type annotation seem off

SciPioneer replied:
+1

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -0,0 +1,236 @@
+    def test_chunked_sharding_basic(self):
+        self.init_pg()
+
+        for dim in [0, -2]:
```
Should `[0, -2]` here be `[0, -1]` or `[0, 1]`? Since the created tensor only has 2 dimensions, dim=-2 actually is equivalent to dim=0.

SciPioneer replied:
Never mind. Saw another test `test_chunked_sharding_columns`.

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -0,0 +1,236 @@
+
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_chunked_sharding_basic(self):
```
Nit: "basic" does not sound like a good descriptive test name. What about "shared_by_alll_4_devices"?

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -0,0 +1,236 @@
+
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_sharded_tensor_partial_world_size(self):
```
Nit: Change "partial_world_size" to "sharded_by_2_of_4_devices"?

SciPioneer replied:
It's up to you. Just feel that `partial_world_size` sounds less straightforward. If you keep this name, you can rename the above test case as "***_full_world_size".

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -97,86 +108,87 @@ def placement(self) -> ShardPlacement:
+        shard_lengths(List[int]): Lengths indicating the length of each
+            dimension for this shard. Should have the same rank as the
+            original tensor.
+        placement(List[Device] or List[PlacementSpec]):
```
`List[Device or PlacementSpec]` is more accurate.

SciPioneer commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -0,0 +1,41 @@
+    ShardedTensor,
+)
+
+def empty(
```
Do we need to consider `layout=torch.strided` here?

pritamdamania87 replied:
I feel `process_grop` should probably be in the end since in most cases I expect you wouldn't specify it.

SciPioneer commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -0,0 +1,41 @@
+def empty(
+        sharding_spec: ShardingSpec,
+        *size,
+        dtype=None,
```
You can add an internal function `init` that requires an `init_meth` besides all the args here, where `init_meth` can be `torch.empty`, `torch.zeros`, or `torch.ones`. This way it will be trivial to support zeroes and ones on the sharded tensor.

pritamdamania87 replied:
I think we can refactor this once we need to add more functionality. Basically in the PR where we add such additional methods, it would be easier to properly refactor this.

SciPioneer replied:
Sure. That's why I also label this comment as optional. Such refactoring can easily be done in the future on demand.

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -0,0 +1,186 @@
+from torch.distributed.utils import _parse_remote_device
+
+
+class ShardWithMetadata(object):
```
Can't we just call this class `Shard`, and this class can have two fields called '_tensor' and `_metadata`? We can also even rename `ShardMetadata` as `Metadata` if there is not other metadata classes in the same package.

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -0,0 +1,186 @@
+
+
+class ShardWithMetadata(object):
+    """
```
Consider `__slot__` for memory optimization?

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -37,7 +37,16 @@ def device(self) -> Device:
+    pass
+
+
+class ChunkShardingSpec(ShardingSpec):
```
I think we should describe clearly what will happen if the # of elements in the specified sharding dimension cannot be divided by the # of shards. After all, the implementation here is different from the data partitioning behavior of 'DataParallel` for the similar case.

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -0,0 +1,236 @@
+                self.assertEqual([shard_rank * 5, 0], shard_metadata.shard_offsets)
+                self.assertEqual([5, 20], shard_metadata.shard_lengths)
+                self.assertEqual(f'rank:{shard_rank + 1}/cuda:{shard_rank + 2}', shard_metadata.placement)
+
```
Can you add another test where the sharding dim only has 2 elements, but you want to shard it using 4 devices?

pritamdamania committed:
Update on "ShardedTensor framework for ChunkedShardingSpec"


Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]

SciPioneer commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -39,12 +39,26 @@ def device(self) -> Device:
     across multiple devices. In particular, it represents sharding a Tensor
     along a single dimension into equal chunks (similar to :meth:`torch.chunk`).
 
+    The semantics of how a tensor is partitioned is inline with
```
Nice comment!

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -0,0 +1,276 @@
+            self.assertEqual(1, len(local_shards))
+            local_shard = local_shards[0].tensor
+            self.assertEqual(torch.device(f"cuda:{self.rank}"), local_shard.device)
+            self.assertEqual((1, 20), local_shard.size())
```
Can we also test the other code path here?

```
else:
    self.assertEqual(0, len(local_shards))
```

SciPioneer commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -0,0 +1,206 @@
+    Container which holds the data for a shard as a Tensor and also
+    the associated metadata for that shard.
+    """
+    __slots__ = ['tensor', 'metadata']
```
It seems that we cannot add `__slots__` here:

ValueError: 'tensor' in __slots__ conflicts with class variable

pritamdamania committed:
Update on "ShardedTensor framework for ChunkedShardingSpec"


Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "ShardedTensor framework for ChunkedShardingSpec"


Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "ShardedTensor framework for ChunkedShardingSpec"


Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "ShardedTensor framework for ChunkedShardingSpec"


Building upon the sharding specifications, this PR introduces the
intial skeleton of ShardedTensor and allows building a ShardedTensor by
specifying ChunkedShardingSpec.

In follow up PRs, I'll add further support for GenericShardingSpec.

Differential Revision: [D28526012](https://our.internmc.facebook.com/intern/diff/D28526012/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28526012/)!

[ghstack-poisoned]
